### PR TITLE
kubelet: support hugepages limit for --system-reserved and --kube-reserved

### PR DESCRIPTION
**Pull request by SataQiu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
kubelet: support hugepages limit for --system-reserved and --kube-reserved

kubelet has the `--reserved-memory` flag, which is described as follows:
```bash
--reserved-memory reserved-memory                          
    A comma separated list of memory reservations for NUMA nodes. (e.g. --reserved-memory 0:memory=1Gi,hugepages-1M=2Gi --reserved-memory 1:memory=2Gi). The total sum for each memory type should be equal to the sum of kube-reserved, system-reserved and eviction-threshold.
```
As described above, `--reserved-memory` supports the setting of hugepages resource reservation, and the reserved value should be equal to the sum of `kube-reserved`, `system-reserved` and eviction-threshold. 
However `--system-reserved` and `--kube-reserved` do not support hugepages yet.
Without this feature, it seems that we cannot reserve some hugepages resources for system daemon.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
kubelet now accepts `hugepages-<size>=<number>` in the `--system-reserved` and `--kube-reserved` options to ensure that the specified number of hugepages will be reserved for the system as a whole and for Kubernetes system daemons respectively. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by ffromani** [comment]
I'm personally in favor of a change like this, but do we have an issue? what's the context?

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -1336,17 +1337,18 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
 	}
 	rl := make(v1.ResourceList)
 	for k, v := range m {
-		switch v1.ResourceName(k) {
-		// CPU, memory, local storage, and PID resources are supported.
-		case v1.ResourceCPU, v1.ResourceMemory, v1.ResourceEphemeralStorage, pidlimit.PIDs:
+		rn := v1.ResourceName(k)
+		switch {
+		// CPU, memory, local storage, PID and hugepages resources are supported.
+		case rn == v1.ResourceCPU, rn == v1.ResourceMemory, rn == v1.ResourceEphemeralStorage, rn == pidlimit.PIDs, v1helper.IsHugePageResourceName(rn):
```
Comment (haircommander) [inline cmd/kubelet/app/server.go:1343]: why the helper func when the others are strict equality checks?

Comment (haircommander) [reply cmd/kubelet/app/server.go:1343]: nevermind, the check can't be equality https://github.com/SataQiu/kubernetes/blob/7b1b8d57340e686d013ee7a4180e04b4b16d4606/pkg/api/v1/resource/helpers.go#L301

File: pkg/kubelet/cm/qos_container_manager_linux.go [hunk 2]
```diff
@@ -159,7 +159,11 @@ func (m *qosContainerManagerImpl) setHugePagesUnbounded(cgroupConfig *CgroupConf
 }
 
 func (m *qosContainerManagerImpl) setHugePagesConfig(configs map[v1.PodQOSClass]*CgroupConfig) error {
-	for _, v := range configs {
+	for qosClass, v := range configs {
+		// the top guaranteed QOS class should respect the reservation for huge pages
```
Comment (SataQiu) [inline pkg/kubelet/cm/qos_container_manager_linux.go:163]: This hugepages limit value is set to the correct values during `enforceNodeAllocatableCgroups`, but qos manager is wrongly to refresh it to maximum.
https://github.com/kubernetes/kubernetes/blob/69b648a1d7074cbe004bf7adb3cdb17f01a4e9d8/pkg/kubelet/cm/node_container_manager_linux.go#L68-L114

File: pkg/kubelet/cm/qos_container_manager_linux.go [hunk 3]
```diff
@@ -159,7 +159,11 @@ func (m *qosContainerManagerImpl) setHugePagesUnbounded(cgroupConfig *CgroupConf
...
+		if qosClass == v1.PodQOSGuaranteed {
+			continue
+		}
```
Comment (SataQiu) [inline pkg/kubelet/cm/qos_container_manager_linux.go:166]: This hugepages limit value is set to the correct values during `enforceNodeAllocatableCgroups`, but qos container manager is wrongly to refresh it to maximum.
https://github.com/kubernetes/kubernetes/blob/69b648a1d7074cbe004bf7adb3cdb17f01a4e9d8/pkg/kubelet/cm/node_container_manager_linux.go#L68-L114

Comment (SataQiu) [inline pkg/kubelet/cm/qos_container_manager_linux.go:166]: This hugepages limit value is set to the correct values during `enforceNodeAllocatableCgroups`, but qos container manager is wrongly to refresh it to maximum.
https://github.com/kubernetes/kubernetes/blob/69b648a1d7074cbe004bf7adb3cdb17f01a4e9d8/pkg/kubelet/cm/node_container_manager_linux.go#L68-L114

**Comment by kannon92** [comment]
> I'm personally in favor of a change like this, but do we have an issue? what's the context? EDIT: TL;DR: do we have use cases for this feature?

Agree to this. I don't think sig-node would be opposed to this change but these usually require some kind of KEP or at least coming to a sig-node meeting to see about the feature.

API changes usually require KEPs and flags for kubelet are considered an API AFAIK.

**Comment by zhonglin6666** [comment]
hugepage unable to map backing store for guest RAM Cannot allocate memory #132084

**Pull request closed without merging**